### PR TITLE
Custom criteria test: particle count

### PR DESCRIPTION
**Pull request by EtherealMC-Bit**
This is a simple test file that shows that the simulation will stop based on user defined criteria. In this case the simulation stops when the total particle count goes above 2.5 million particles.

### CODE REVIEW

**Review by roelof-groenewald — COMMENTED** [review]
@MKieburtz How long does this test take to run on your machine? We should try to keep tests quick, so with that thought maybe decrease the total particle count by an order of magnitude or 2. The test will be just as effective with a smaller number of total particles but it might run significantly faster.

File: mewarpx/mewarpx/diags_store/diag_base.py [hunk 1]
```diff
@@ -3,7 +3,7 @@
 import time
 
 import numpy as np
-import psutil
+#import psutil
```
Comment (peterscherpelz) [inline mewarpx/mewarpx/diags_store/diag_base.py]: Add a space after `#` please.

File: mewarpx/tests/test_custom_criteria.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    # TODO: what can be kept here between these two lines?
+    # these are commented out because I don't know what to pass in for capsys
+    # run.runresults.finalize_save()
+    # out, _ = capsys.readouterr()
+
+    # print(out)
+    # # make sure out isn't empty
+    # assert out
+# this is how I'm running it
+test_capacitive_discharge_multigrid(None, "Run2D")
```
Comment (peterscherpelz) [inline mewarpx/tests/test_custom_criteria.py]: @MKieburtz I'm guessing this is your problem right now. `pytest` automatically runs functions that begin with `test`, and in general modules should *not* execute code like this when they're just imported. Sorry, I fixed it last minute in the main diode template and forgot to point it out to you, but see everything below line 84 https://github.com/ModernElectron/WarpX/blob/b2a85ee6fb43e4a58a4f616388f5e74f33c118ae/mewarpx/tests/test_capacitive_discharge.py#L84 - the `if __name__` block is what solves your problem here; it runs the test if you call it with `python` but not if you call it with `pytest`.

**Comment by EtherealMC-Bit** [comment]
> How long does this test take to run on your machine?

@roelof-groenewald it takes about 4 seconds on penning. I'll go ahead and lower the number of particles.

File: mewarpx/tests/test_custom_criteria.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    # Initialize and import only when we know dimension
+    mwxutil.init_libwarpx(ndim=dim, rz=use_rz)
+    from mewarpx import testing_util
+    from mewarpx.setups_store import diode_setup
+    from mewarpx import mwxrun
+    from pywarpx import callbacks
+
+    # Include a random run number to allow parallel runs to not collide. Using
+    # python randint prevents collisions due to numpy rseed below
+    #testing_util.initialize_testingdir(name)
```
Comment (peterscherpelz) [inline mewarpx/tests/test_custom_criteria.py]: You do want to be using this function

File: mewarpx/tests/test_custom_criteria.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        D_CA=D_CA,
+        INERT_GAS_TYPE='He',
+        N_INERT=9.64e20,  # m^-3
+        T_INERT=300.0,  # K
+        PLASMA_DENSITY=2.56e14,  # m^-3
+        T_ELEC=30000.0,  # K
+        NX=16,
+        NZ=128,
+        # This gives equal spacing in x & z
+        PERIOD=D_CA * 16 / 128.0,
```
Comment (peterscherpelz) [inline mewarpx/tests/test_custom_criteria.py:68]: See NX and NY predefined at https://github.com/ModernElectron/WarpX/blob/b2a85ee6fb43e4a58a4f616388f5e74f33c118ae/mewarpx/tests/test_capacitive_discharge.py#L45 please use that.

In general please do a diff against the `test_capacitive_discharge` in `memaster` to make sure you pick up all changes from there.

File: mewarpx/tests/test_custom_criteria.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        DT=DT,
+        TOTAL_TIMESTEPS=10,
+        DIAG_STEPS=DIAG_STEPS,
+        DIAG_INTERVAL=DIAG_INTERVAL,
+        NUMBER_PARTICLES_PER_CELL=[16, 32],
+        FIELD_DIAG_DATA_LIST=['rho_electrons', 'rho_he_ions', 'phi'],
+        SIMULATION=sim
```
Comment (peterscherpelz) [inline mewarpx/tests/test_custom_criteria.py]: Another change in newer test

File: mewarpx/tests/test_custom_criteria.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        init_simcontrol=True,
+        init_warpx=True
+    )
+    run.control.add_checker(check_particle_nums)
+
+    callbacks.installafterstep(add_particles)
+
+    # Run the main WARP loop
+    while run.control.check_criteria():
+        run.SIMULATION.step(1)
```
Comment (peterscherpelz) [inline mewarpx/tests/test_custom_criteria.py]: Another change in newer test

File: mewarpx/tests/test_custom_criteria.py [hunk 7]
```diff
@@ -90,19 +89,22 @@ def test_capacitive_discharge_multigrid(capsys, name):
...
+
 
-    # print(out)
-    # # make sure out isn't empty
-    # assert out
-# this is how I'm running it
-test_capacitive_discharge_multigrid(None, "Run2D")
\ No newline at end of file
+if __name__ == '__main__':
+    test_capacitive_discharge_custom_criteria(None, "Run2D")
```
Comment (peterscherpelz) [inline mewarpx/tests/test_custom_criteria.py]: Newline here please.

**Review by peterscherpelz — APPROVED** [review]
@MKieburtz I'll mark this approved on my part, though please do add the end-of-file newline on the test. Thanks for all the work!

**Review by KZhu-ME — APPROVED** [review]
Looks good!

**Pull request merged**